hs_labels/models.py

```diff
@@ -199,113 +233,157 @@ def clean_label(name):
         label_string = UserLabels.clean_label(this_label)
-        UserResourceLabels.objects.get_or_create(resource=this_resource,
-                                                 label=label_string,
-                                                 user=self.user)
+        with transaction.atomic():  # empirically, get_or_create is not atomic.
+            UserResourceLabels.objects.get_or_create(resource=this_resource,
+                                                     label=label_string,
+                                                     user=self.user)
 
     def unlabel_resource(self, this_resource, this_label):
-        """ Remove one label from a resource
+        """
+        Remove one label from a resource
 
-        :param this_resource: BaseResource from which to remove a label
-        :param this_label: The label to remove. Only that labels will be removed.
-        :return: none
+        Users are allowed to label any resource, including resources to which they do not have access.
+        This is not an access control problem because labeling information is private.
         """
+        if __debug__:  # during testing only, check argument types and preconditions
+            assert isinstance(this_resource, BaseResource)
+            assert isinstance(this_label, basestring)
 
         # remove leading and trailing spaces
         label_string = UserLabels.clean_label(this_label)
+
         UserResourceLabels.objects.filter(resource=this_resource,
                                           label__exact=label_string,
                                           user=self.user).delete()
 
     def clear_resource_labels(self, this_resource):
-        """ Clear all labels for a resource
-
-        :param this_resource: BaseResource for which to clear all labels
-        :return: none.
         """
+        Clear all labels for a resource
+        """
+        if __debug__:  # during testing only, check argument types and preconditions
+            assert isinstance(this_resource, BaseResource)
 
         UserResourceLabels.objects.filter(resource=this_resource,
                                           user=self.user).delete()
 
     def remove_resource_label(self, this_label):
         """
-        clear the specified label from all assigned resources
+        clear a label from the labeling system.
         """
-        UserResourceLabels.objects.filter(label__exact=this_label, user=self.user).delete()
+        if __debug__:  # during testing only, check argument types and preconditions
+            assert isinstance(this_label, basestring)
+        UserResourceLabels.objects.filter(label=this_label, user=self.user)\
+                                  .delete()
 
-    def favorite_resource(self, this_resource):
+    ##########################################
+    # general flagging of resources
+    ##########################################
+
+    def flag_resource(self, this_resource, this_flagcode):
```

Comment:
Should this helper function be private? Otherwise, this should have corresponding unit tests.